File: reference/internal/ui/metrics/labels.go
package metrics

import (
	"strings"

	"github.com/Elpulgo/azdo/internal/config"
)

// stateLabels resolves the three column-header labels from the model's
// config. Order of precedence per state: explicit override in
// `metrics.state_labels`, then auto-derived from the configured state name.
//
// Returns (active, readyForTest, closed) labels.
func (m Model) stateLabels() (string, string, string) {
	sc := m.stateConfig()
	var lbl config.MetricsStates
	if m.config != nil {
		lbl = m.config.Metrics.StateLabels
	}
	return labelFor(sc.Active, lbl.Active),
		labelFor(sc.ReadyForTest, lbl.ReadyForTest),
		labelFor(sc.Closed, lbl.Closed)
}

// labelFor returns the override if non-empty, else an auto-derived
// abbreviation from `name`. The abbreviation rule:
//   - multi-word names → initials, lowercase ("Ready for Test" → "rft",
//     "In Progress" → "ip")
//   - single-word names → lowercase as-is ("Active" → "active",
//     "Done" → "done")
func labelFor(name, override string) string {
	if strings.TrimSpace(override) != "" {
		return override
	}
	words := strings.Fields(strings.TrimSpace(name))
	if len(words) == 0 {
		return ""
	}
	if len(words) == 1 {
		return strings.ToLower(words[0])
	}
	var b strings.Builder
	for _, w := range words {
		b.WriteByte(strings.ToLower(w)[0])
	}
	return b.String()
}

// labelTitle returns the same label as labelFor but with the first letter
// upper-cased so column headers feel like proper titles ("Active",
// "RFT" → "Rft"; the override survives intact if the user wrote "RFT").
func labelTitle(name, override string) string {
	if strings.TrimSpace(override) != "" {
		return override
	}
	return titleCase(labelFor(name, ""))
}

// titleCase upper-cases the first byte of an ASCII string. Safe for the
// short labels we deal with here (state names, abbreviations).
func titleCase(s string) string {
	if s == "" {
		return ""
	}
	return strings.ToUpper(s[:1]) + s[1:]
}

File: src/Azdo.Tui/Views/Metrics/Labels.cs
namespace Azdo.Tui.Views.Metrics;

/// <summary>
/// State-label helpers: resolve the three column-header labels from the
/// configured state names, honoring explicit <c>metrics.state_labels</c>
/// overrides.
/// </summary>
public static class Labels
{
    /// <summary>
    /// The override if non-empty, else an auto-derived abbreviation from
    /// <paramref name="name"/>:
    /// multi-word names → lowercase initials ("Ready for Test" → "rft");
    /// single-word names → lowercase as-is ("Active" → "active").
    /// </summary>
    public static string LabelFor(string name, string @override)
    {
        if (!string.IsNullOrWhiteSpace(@override))
            return @override;
        var words = (name ?? "").Trim().Split((char[]?)null, StringSplitOptions.RemoveEmptyEntries);
        if (words.Length == 0)
            return "";
        if (words.Length == 1)
            return words[0].ToLowerInvariant();
        var chars = new char[words.Length];
        for (int i = 0; i < words.Length; i++)
            chars[i] = char.ToLowerInvariant(words[i][0]);
        return new string(chars);
    }

    /// <summary>
    /// Same as <see cref="LabelFor"/> but with the first letter upper-cased so
    /// column headers feel like proper titles. The override survives intact.
    /// </summary>
    public static string LabelTitle(string name, string @override)
    {
        if (!string.IsNullOrWhiteSpace(@override))
            return @override;
        return TitleCase(LabelFor(name, ""));
    }

    /// <summary>Upper-cases the first character of an ASCII string.</summary>
    public static string TitleCase(string s)
    {
        if (string.IsNullOrEmpty(s))
            return "";
        return char.ToUpperInvariant(s[0]) + s[1..];
    }
}
